File: src/gaet/detect.py
"""Local PostgreSQL instance detection (socket + TCP fallback)."""

import os
from typing import Dict, List
from .core import Dict, List, _find_socket_paths, _socket_port, getpass, os, run_cmd

def detect_local_pg(psql_path: str) -> List[Dict[str, str]]:
    """
    Auto-detect running PostgreSQL instances on this machine.
    Returns list of dicts with keys: host, port, user, databases.
    Detects both TCP (127.0.0.1:port) and Unix socket connections.
    """
    results: List[Dict[str, str]] = []
    if not psql_path:
        return results

    # Users to try: current OS user first (peer auth: OS user = DB role),
    # then the conventional postgres/root accounts.
    users_to_try = ["postgres", "root"]
    try:
        import getpass
        cur = getpass.getuser()
        if cur and cur not in users_to_try:
            users_to_try.insert(0, cur)
    except Exception:
        pass

    # --- 1. Try Unix sockets first (common on Linux) ---
    # Scan every .s.PGSQL.* file so non-default ports (5433, 5434, ...)
    # are detected too, and keep going after the first hit so multiple
    # instances on different sockets are all reported.
    seen_ports: set = set()
    for sock in _find_socket_paths():
        port = _socket_port(sock)
        if port in seen_ports:
            continue  # same port already found (multi-dir sockets)
        host = os.path.dirname(sock)
        for user in users_to_try:
            out, _, rc = run_cmd(
                [psql_path, "-w", "-h", host, "-p", port, "-U", user,
                 "-d", "postgres", "-tAc", "SELECT current_database();"],
                env={"PGPASSWORD": ""}, timeout=3,
            )
            if rc == 0 and out.strip():
                db = out.strip()
                # List all databases on this server
                dbs_out, _, _ = run_cmd(
                    [psql_path, "-w", "-h", host, "-p", port, "-U", user,
                     "-d", "postgres", "-tAc",
                     "SELECT datname FROM pg_database WHERE datistemplate = false ORDER BY datname;"],
                    env={"PGPASSWORD": ""}, timeout=3,
                )
                databases = [d.strip() for d in dbs_out.strip().split("\n") if d.strip()] if dbs_out.strip() else [db]
                results.append({
                    "host": host,  # socket directory
                    "port": port,
                    "user": user,
                    "databases": ", ".join(databases),
                    "default_db": db,
                })
                seen_ports.add(port)
                break  # Found a working user on this socket

    # --- 2. Try TCP ports (fallback) — skip ports found via socket so an
    # instance that answered on its socket is not reported a second time.
    ports_to_try = ["5432", "5433", "5434", "5435", "5436"]

    for port in ports_to_try:
        if port in seen_ports:
            continue  # already found via Unix socket
        for user in users_to_try:
            # Try connecting with no password (common for local dev)
            out, _, rc = run_cmd(
                [psql_path, "-w", "-h", "127.0.0.1", "-p", port, "-U", user,
                 "-d", "postgres", "-tAc",
                 "SELECT current_database();"],
                env={"PGPASSWORD": ""},
                timeout=3,
            )
            if rc == 0 and out.strip():
                db = out.strip()
                # List all databases on this server
                dbs_out, _, _ = run_cmd(
                    [psql_path, "-w", "-h", "127.0.0.1", "-p", port, "-U", user,
                     "-d", "postgres", "-tAc",
                     "SELECT datname FROM pg_database WHERE datistemplate = false ORDER BY datname;"],
                    env={"PGPASSWORD": ""}, timeout=3,
                )
                databases = [d.strip() for d in dbs_out.strip().split("\n") if d.strip()] if dbs_out.strip() else [db]
                results.append({
                    "host": "127.0.0.1",
                    "port": port,
                    "user": user,
                    "databases": ", ".join(databases),
                    "default_db": db,
                })
                break  # Found this port, no need to try other users

    return results

File: pkg/detect/detect.go
// Package detect provides local PostgreSQL auto-discovery via Unix socket and TCP probing.
package detect

import (
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/ghanirahmans/gaet/pkg/core"
)

// DBInstance holds information about a discovered database instance.
type DBInstance = PGInstance

// DetectLocalDB discovers running database instances.
func DetectLocalDB(psqlPath string) []DBInstance {
	return DetectLocalPG(psqlPath)
}

// PGInstance holds information about a discovered PostgreSQL instance.
type PGInstance struct {
	Host      string   `json:"host"`
	Port      string   `json:"port"`
	User      string   `json:"user"`
	Databases []string `json:"databases"`
	DefaultDB string   `json:"default_db"`
	IsSocket  bool     `json:"is_socket"`
}

// DetectLocalPG discovers running PostgreSQL instances (socket first, then TCP fallback).
func DetectLocalPG(psqlPath string) []PGInstance {
	if psqlPath == "" {
		return nil
	}

	var results []PGInstance
	seenPorts := map[string]bool{}

	usersToTry := buildUserList()

	// 1. Unix socket scan
	for _, sockPath := range core.FindSocketPaths() {
		port := core.SocketPort(sockPath)
		if seenPorts[port] {
			continue
		}
		host := filepath.Dir(sockPath)
		for _, user := range usersToTry {
			stdout, _, rc := core.RunCmdSimple(psqlPath,
				[]string{"-w", "-h", host, "-p", port, "-U", user, "-d", "postgres", "-tAc", "SELECT current_database();"},
				map[string]string{"PGPASSWORD": ""}, 3*time.Second)
			if rc == 0 && strings.TrimSpace(stdout) != "" {
				db := strings.TrimSpace(stdout)
				dbs := listDatabases(psqlPath, host, port, user)
				if len(dbs) == 0 {
					dbs = []string{db}
				}
				results = append(results, PGInstance{
					Host:      host,
					Port:      port,
					User:      user,
					Databases: dbs,
					DefaultDB: db,
					IsSocket:  true,
				})
				seenPorts[port] = true
				break
			}
		}
	}

	// 2. TCP ports fallback
	for _, port := range []string{"5432", "5433", "5434", "5435", "5436"} {
		if seenPorts[port] {
			continue
		}
		for _, user := range usersToTry {
			stdout, _, rc := core.RunCmdSimple(psqlPath,
				[]string{"-w", "-h", "127.0.0.1", "-p", port, "-U", user, "-d", "postgres", "-tAc", "SELECT current_database();"},
				map[string]string{"PGPASSWORD": ""}, 3*time.Second)
			if rc == 0 && strings.TrimSpace(stdout) != "" {
				db := strings.TrimSpace(stdout)
				dbs := listDatabasesTCP(psqlPath, "127.0.0.1", port, user)
				if len(dbs) == 0 {
					dbs = []string{db}
				}
				results = append(results, PGInstance{
					Host:      "127.0.0.1",
					Port:      port,
					User:      user,
					Databases: dbs,
					DefaultDB: db,
					IsSocket:  false,
				})
				break
			}
		}
	}

	return results
}

func buildUserList() []string {
	users := []string{"postgres", "root"}
	if cur, err := currentUser(); err == nil && cur != "" {
		found := false
		for _, u := range users {
			if u == cur {
				found = true
				break
			}
		}
		if !found {
			users = append([]string{cur}, users...)
		}
	}
	return users
}

func currentUser() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	// Simplest approach: use $USER env
	if u := os.Getenv("USER"); u != "" {
		return u, nil
	}
	// Fallback: parse home dir basename
	return filepath.Base(home), nil
}

func listDatabases(psqlPath, host, port, user string) []string {
	stdout, _, rc := core.RunCmdSimple(psqlPath,
		[]string{"-w", "-h", host, "-p", port, "-U", user, "-d", "postgres", "-tAc",
			"SELECT datname FROM pg_database WHERE datistemplate = false ORDER BY datname;"},
		map[string]string{"PGPASSWORD": ""}, 3*time.Second)
	if rc != 0 || strings.TrimSpace(stdout) == "" {
		return nil
	}
	var dbs []string
	for _, line := range strings.Split(strings.TrimSpace(stdout), "\n") {
		d := strings.TrimSpace(line)
		if d != "" {
			dbs = append(dbs, d)
		}
	}
	return dbs
}

func listDatabasesTCP(psqlPath, host, port, user string) []string {
	return listDatabases(psqlPath, host, port, user)
}
